File: 52-n-queens-ii/n-queens-ii.java
class Solution {
    int resultCount = 0;

    public int totalNQueens(int n) {
        int[] placement = new int[n];

        Arrays.fill(placement, -1);

        placeQueens(0, n, placement);

        return resultCount;
    }

    private void placeQueens(int row, int n, int[] placement) {
        if (row == n) {
            resultCount++;
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isValidPlacement(row, col, placement)) {
                placement[row] = col;

                placeQueens(row + 1, n, placement);

                placement[row] = -1;
            }
        }
    }

    private boolean isValidPlacement(int row, int col, int[] placement) {
        for (int i = 0; i < row; i++) {
            if (placement[i] == col || Math.abs(placement[i] - col) == row - i) {
                return false;
            }
        }
        return true;
    }

}

File: 52-n-queens-ii/n-queens-ii.py
class Solution(object):
    def __init__(self):
        self.result_count = 0

    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        placement = [-1] * n
        self.place_queens(0, n, placement)
        return self.result_count

    def place_queens(self, row, n, placement):
        if row == n:
            self.result_count += 1
            return

        for col in range(n):
            if self.is_valid_placement(row, col, placement):
                placement[row] = col
                self.place_queens(row + 1, n, placement)
                placement[row] = -1

    def is_valid_placement(self, row, col, placement):
        for i in range(row):
            if placement[i] == col or abs(placement[i] - col) == row - i:
                return False
        return True